File: scorio/eval/max_reward.py
r"""Max-reward metrics and uncertainty estimators for categorical outcomes.

``Max@k`` is the continuous-reward generalization of Pass@k: instead of asking
whether at least one sampled response is correct, it scores the best response
among ``k`` sampled traces according to a user-specified reward scale.

The point estimator implemented here matches the appendix evaluation formula in
Bagirov et al. (2025), "The Best of N Worlds: Aligning Reinforcement Learning
with Best-of-N Sampling via max@k Optimization" (Appendix C.1 / Listing 1,
arXiv:2510.23393, https://arxiv.org/abs/2510.23393). This module adapts that
estimator to ``scorio``'s categorical outcome representation via ``R`` plus a
reward map ``w``.

The companion ``*_ci`` functions are a ``scorio`` Bayesian extension. They use
the same grouped-Dirichlet posterior model as :func:`~scorio.eval.bayes` and
are not part of the paper above.

Methods
-------
- ``max_at_k``: expected best reward among ``k`` sampled traces.

Each metric has a companion ``*_ci`` function that returns
``(mu, sigma, lo, hi)`` under the Bayesian uncertainty model used here.
"""

import math

import numpy as np
from scipy.stats import hypergeom

from ._categorical import (
    CategoricalBank,
    _scaled_rewards,
    observed_mean,
    prepare_categorical_bank,
)
from ._inputs import validate_finite_k, validate_latent_k
from .bayes import bayes_ci
from .utils import normal_credible_interval


def _hypergeom_any_above(
    cumulative_counts: np.ndarray,
    trial_count: int,
    k: int,
) -> np.ndarray:
    """Probability that at least one draw exceeds a reward threshold."""
    above_counts = trial_count - cumulative_counts
    probabilities = hypergeom.sf(0, trial_count, above_counts, k)
    return np.asarray(probabilities, dtype=float)


def _log_beta_power_moment(alpha: float, beta: float, power: int) -> float:
    """Return ``log(E[p**power])`` without subtracting large log-gammas."""
    if power == 0:
        return 0.0

    offsets = np.arange(power, dtype=float)
    denominators = alpha + beta + offsets
    complement = beta / denominators
    terms = np.empty(power, dtype=float)
    near_one = complement <= 0.5
    terms[near_one] = np.log1p(-complement[near_one])
    terms[~near_one] = np.log(alpha + offsets[~near_one]) - np.log(
        denominators[~near_one]
    )
    return math.fsum(float(term) for term in terms)


def _log_dirichlet_nested_cumulative_moment(
    total: float, a: float, b: float, k: int
) -> float:
    """Return ``log(E[X^k (X+Y)^k])`` for a Dirichlet partition.

    Writing ``U = X + Y`` and ``V = X / (X + Y)``, Dirichlet neutrality gives
    independent ``U ~ Beta(a+b, total-a-b)`` and ``V ~ Beta(a, b)``. Hence
    ``E[X^k U^k] = E[U^(2k)] E[V^k]``.
    """
    if b <= 0.0:
        raise ValueError("b must be > 0 for nested cumulative moments")
    return _log_beta_power_moment(
        a + b,
        total - a - b,
        2 * k,
    ) + _log_beta_power_moment(a, b, k)


def _covariance_from_log_moments(
    log_cross_moment: float,
    log_left_mean: float,
    log_right_mean: float,
) -> float:
    """Subtract two positive moments without cancellation."""
    log_product = log_left_mean + log_right_mean
    difference = log_product - log_cross_moment
    tolerance = (
        64.0
        * np.finfo(float).eps
        * max(
            1.0,
            abs(log_product),
            abs(log_cross_moment),
        )
    )
    if difference > 0.0:
        if difference <= tolerance:
            difference = 0.0
        else:
            raise FloatingPointError(
                "nested cumulative covariance is materially negative"
            )
    return math.exp(log_cross_moment) * -math.expm1(difference)


def max_at_k(R: np.ndarray, k: int, w: np.ndarray | None = None) -> float:
    r"""
    Max@k: expected best reward among ``k`` sampled traces.

    When ``w = [0, 1]``, Max@k reduces exactly to Pass@k. More generally, the
    reward vector ``w`` maps categorical outcomes to arbitrary real-valued
    scores, and Max@k averages the best score obtainable from a subset of size
    ``k``.

    References:
        - Bagirov, F., et al. (2025). The Best of N Worlds: Aligning
          Reinforcement Learning with Best-of-N Sampling via max@k
          Optimization. *arXiv:2510.23393*.
          https://arxiv.org/abs/2510.23393
          The finite-sample estimator below matches Appendix C.1 / Listing 1.
        - Walder, C., & Karkhanis, D. (2025). Pass@K Policy Optimization:
          Solving Harder Reinforcement Learning Problems. *arXiv:2505.15201*.

    Args:
        R: :math:`M \times N` categorical outcome matrix with integer entries
           in :math:`\{0, \ldots, C\}`.
        k: Number of selected samples, with ``1 <= k <= N``.
        w: Optional reward vector of shape ``(C+1,)``. If omitted, ``R`` must
           be binary and ``w = [0, 1]`` is used.

    Returns:
        float: Average Max@k score across prompts.

    Formula:
        Let :math:`g_{\alpha 1} \le \cdots \le g_{\alpha N}` denote the
        sorted rewards for prompt :math:`\alpha`. Then the unbiased finite-
        sample estimator is

        .. math::

            \mathrm{Max@k}_\alpha = \frac{1}{\binom{N}{k}}
            \sum_{i=k}^{N} \binom{i-1}{k-1} g_{\alpha i}.

        The dataset-level metric is the average across prompts:

        .. math::

            \mathrm{Max@k} = \frac{1}{M}
            \sum_{\alpha=1}^{M} \mathrm{Max@k}_\alpha

    Examples:
        Binary (reduces to Pass@k):

        >>> import numpy as np
        >>> R = np.array([[0, 1, 1, 0, 1],
        ...               [1, 1, 0, 1, 1]])
        >>> round(max_at_k(R, 2), 6)
        0.95

        Weighted categorical rewards:

        >>> R = np.array([[0, 1, 2, 2, 1],
        ...               [1, 1, 0, 2, 2]])
        >>> w = np.array([0.0, 0.5, 1.0])
        >>> round(max_at_k(R, 2, w=w), 6)
        0.85
    """
    bank = prepare_categorical_bank(R, w=w)
    k = validate_finite_k(bank.trial_count, k)
    if k == 1:
        return observed_mean(bank)

    grouped_counts, levels = bank.grouped_observed_counts()
    if levels.size == 1:
        return float(levels[0])

    cumulative_counts = np.cumsum(grouped_counts, axis=1)[:, :-1]
    any_above = _hypergeom_any_above(
        cumulative_counts,
        bank.trial_count,
        k,
    )
    offset, reward_scale, normalized_levels = _scaled_rewards(levels)
    normalized_scores = normalized_levels[0] + any_above @ np.diff(normalized_levels)
    return float(offset + reward_scale * np.mean(normalized_scores))


def _max_at_k_bayes(
    bank: CategoricalBank,
    k: int,
) -> tuple[float, float]:
    """Posterior mean/std for Max@k under a grouped Dirichlet posterior."""
    gamma, levels = bank.grouped_posterior_counts()
    M = bank.question_count
    L = int(levels.size)
    total = float(bank.category_count + bank.prior_trial_count + bank.trial_count)

    # The posterior moments describe the latent distribution, so k is not
    # restricted by the observed sample size once the posterior is defined.

    if L == 1:
        mu = float(levels[0])
        return mu, 0.0

    offset, reward_scale, normalized_levels = _scaled_rewards(levels)
    gaps = np.diff(normalized_levels)

    means = np.empty(M, dtype=float)
    row_sigmas = np.empty(M, dtype=float)

    for row in range(M):
        gamma_row = gamma[row]
        cum = np.cumsum(gamma_row)[:-1]  # A_l parameters for l = 1..L-1

        log_e_ak = np.empty(L - 1, dtype=float)
        log_e_a2k = np.empty(L - 1, dtype=float)
        for idx in range(L - 1):
            a = float(cum[idx])
            b = total - a
            log_e_ak[idx] = _log_beta_power_moment(a, b, k)
            log_e_a2k[idx] = _log_beta_power_moment(a, b, 2 * k)

        exceedance_probabilities = -np.expm1(log_e_ak)
        mean_increment = float(np.dot(gaps, exceedance_probabilities))

        covariance = np.empty((L - 1, L - 1), dtype=float)
        for i in range(L - 1):
            covariance[i, i] = _covariance_from_log_moments(
                log_e_a2k[i],
                log_e_ak[i],
                log_e_ak[i],
            )
            for j in range(i + 1, L - 1):
                a = float(cum[i])
                b = float(cum[j] - cum[i])
                log_cross = _log_dirichlet_nested_cumulative_moment(
                    total,
                    a,
                    b,
                    k,
                )
                value = _covariance_from_log_moments(
                    log_cross,
                    log_e_ak[i],
                    log_e_ak[j],
                )
                covariance[i, j] = value
                covariance[j, i] = value

        normalized_variance = float(gaps @ covariance @ gaps)
        if normalized_variance < 0.0:
            tolerance = 64.0 * np.finfo(float).eps
            if normalized_variance >= -tolerance:
                normalized_variance = 0.0
            else:
                raise FloatingPointError(
                    "Max@k posterior variance is materially negative"
                )

        normalized_mean = float(
            np.clip(
                normalized_levels[0] + mean_increment,
                normalized_levels[0],
                normalized_levels[-1],
            )
        )
        means[row] = offset + reward_scale * normalized_mean
        row_sigmas[row] = reward_scale * math.sqrt(normalized_variance)

    mu = float(np.mean(means))
    sigma_scale = float(np.max(row_sigmas))
    if sigma_scale == 0.0:
        sigma = 0.0
    else:
        sigma = float(
            sigma_scale * math.sqrt(float(np.sum((row_sigmas / sigma_scale) ** 2))) / M
        )
    return mu, sigma


def max_at_k_ci(
    R: np.ndarray,
    k: int,
    w: np.ndarray | None = None,
    R0: np.ndarray | None = None,
    confidence: float = 0.95,
    bounds: tuple[float, float] | None = None,
) -> tuple[float, float, float, float]:
    r"""
    Bayesian posterior summary for :func:`max_at_k`.

    The posterior uses the same Dirichlet-plus-one construction as
    :func:`~scorio.eval.bayes`. When ``k = 1``, ``Max@1`` reduces to the usual
    single-draw expected score, so this function agrees with
    :func:`~scorio.eval.bayes_ci`.

    This uncertainty model is a ``scorio`` extension. Bagirov et al. (2025)
    define the finite-sample max@k point estimator, but do not derive these
    Bayesian credible intervals.

    Args:
        R: :math:`M \times N` categorical outcome matrix with integer entries
           in :math:`\{0, \ldots, C\}`.
        k: Selection count. The posterior target is defined for any integer
           ``k >= 1``; ``k = 1`` matches :func:`~scorio.eval.bayes_ci`.
        w: Optional reward vector of shape ``(C+1,)``. If omitted, ``R`` must
           be binary and ``w = [0, 1]`` is used.
        R0: Optional :math:`M \times D` matrix of prior outcomes.
        confidence: Credibility level for the normal-approximation interval.
        bounds: Optional ``(lo, hi)`` clipping bounds. If omitted, the interval
            is clipped to the minimum and maximum reward levels in ``w``.

    Returns:
        tuple[float, float, float, float]:
            :math:`(\mu,\; \sigma,\; \text{lo},\; \text{hi})`

    Formula:
        Let :math:`r_1 < \cdots < r_L` be the unique reward levels and
        :math:`A_{\alpha \ell}` the posterior cumulative probability of
        obtaining reward at most :math:`r_\ell` for prompt :math:`\alpha`.
        Then the per-prompt latent target is

        .. math::

            g_\alpha = r_L - \sum_{\ell=1}^{L-1}
            (r_{\ell+1} - r_\ell) A_{\alpha \ell}^k

        and posterior moments are computed in closed form under the grouped
        Dirichlet posterior.

    Examples:
        Binary:

        >>> import numpy as np
        >>> R = np.array([[0, 1, 1, 0, 1],
        ...               [1, 1, 0, 1, 1]])
        >>> mu, sigma, lo, hi = max_at_k_ci(R, 2)
        >>> round(mu, 6), round(sigma, 6), round(lo, 4), round(hi, 4)
        (0.839286, 0.097263, 0.6487, 1.0)

        Weighted categorical rewards:

        >>> R = np.array([[0, 1, 2, 2, 1],
        ...               [1, 1, 0, 2, 2]])
        >>> w = np.array([0.0, 0.5, 1.0])
        >>> mu, sigma, lo, hi = max_at_k_ci(R, 2, w=w)
        >>> round(mu, 6), round(sigma, 6), round(lo, 4), round(hi, 4)
        (0.75, 0.08812, 0.5773, 0.9227)
    """
    k = validate_latent_k(k)
    if k == 1:
        return bayes_ci(R, w=w, R0=R0, confidence=confidence, bounds=bounds)

    bank = prepare_categorical_bank(R, w=w, R0=R0)
    mu, sigma = _max_at_k_bayes(bank, k)
    if bounds is None:
        bounds = (float(np.min(bank.weights)), float(np.max(bank.weights)))
    lo, hi = normal_credible_interval(
        mu, sigma, credibility=confidence, two_sided=True, bounds=bounds
    )
    return float(mu), float(sigma), float(lo), float(hi)


__all__ = [
    "max_at_k",
    "max_at_k_ci",
]

File: js/scorio/src/eval/maxReward.ts
/**
 * Max@k family — the continuous-reward generalization of Pass@k. Instead of
 * asking whether at least one sampled response is correct, Max@k scores the
 * best response among `k` sampled traces under a user-specified reward scale.
 * Port of `scorio/eval/max_reward.py`.
 *
 * The point estimator matches the appendix evaluation formula in Bagirov et
 * al. (2025), "The Best of N Worlds: Aligning Reinforcement Learning with
 * Best-of-N Sampling via max@k Optimization" (Appendix C.1 / Listing 1,
 * arXiv:2510.23393). The companion `*_ci` function is a `scorio` Bayesian
 * extension using the same grouped-Dirichlet posterior as `bayes`.
 */

import { gammaln, comb, betaRatio } from "./internal/math.js";
import { normalCredibleInterval, type Bounds } from "./internal/ci.js";
import { asMatrix, validateMatrixRange, type Matrix } from "./internal/validate.js";
import { bayesCi } from "./bayes.js";

/** Numerically stable `log(sum(exp(values)))`, mirroring `scipy.special.logsumexp`. */
function logSumExp(values: readonly number[]): number {
  let max = -Infinity;
  for (const v of values) if (v > max) max = v;
  if (max === -Infinity) return -Infinity;
  let sum = 0;
  for (const v of values) sum += Math.exp(v - max);
  return max + Math.log(sum);
}

/** Per-row counts of the values `0..length-1` over a possibly-empty matrix. */
function rowBincountWide(
  A: readonly (readonly number[])[],
  length: number,
): number[][] {
  return A.map((row) => {
    const counts = new Array<number>(length).fill(0);
    for (const v of row) counts[v]! += 1;
    return counts;
  });
}

/** Normalized `R`, weight vector `w`, and prior matrix `R0`. */
interface CategoricalInput {
  Rm: number[][];
  wv: number[];
  R0m: number[][];
}

/** Normalize R, w, and R0 for weighted categorical metrics. */
function prepareCategoricalInput(
  R: Matrix,
  w?: readonly number[],
  R0?: Matrix,
): CategoricalInput {
  const Rm = asMatrix(R);

  let wv: number[];
  if (w === undefined) {
    const seen = new Set<number>();
    for (const row of Rm) for (const v of row) seen.add(v);
    const isBinary = seen.size <= 2 && [...seen].every((v) => v === 0 || v === 1);
    if (!isBinary) {
      const vals = [...seen].sort((a, b) => a - b).join(", ");
      throw new Error(
        `R contains more than 2 unique values (${vals}), so weight vector 'w' must be provided. ` +
          `Please specify a weight vector of length ${seen.size} to map each category to a score.`,
      );
    }
    wv = [0.0, 1.0];
  } else {
    wv = w.map(Number);
  }

  const M = Rm.length;
  const C = wv.length - 1;
  validateMatrixRange(Rm, 0, C, "R");

  let R0m: number[][];
  if (R0 === undefined) {
    R0m = Rm.map(() => []);
  } else {
    R0m = asMatrix(R0);
    if (R0m.length !== M) {
      throw new Error("R0 must have the same number of rows (M) as R.");
    }
    validateMatrixRange(R0m, 0, C, "R0");
  }

  return { Rm, wv, R0m };
}

function validateK(N: number, k: number): void {
  if (!(k >= 1 && k <= N) || !Number.isInteger(k)) {
    throw new Error(`k must satisfy 1 <= k <= N (N=${N}); got k=${k}`);
  }
}

/** Sorted unique values of `w`, with each entry's index into that sorted list. */
function uniqueLevels(wv: readonly number[]): { levels: number[]; inverse: number[] } {
  const sorted = [...new Set(wv)].sort((a, b) => a - b);
  const index = new Map<number, number>();
  sorted.forEach((v, i) => index.set(v, i));
  return { levels: sorted, inverse: wv.map((v) => index.get(v)!) };
}

/** Grouped Dirichlet posterior parameters and the unique reward levels. */
function groupedPosteriorParams(
  R: Matrix,
  w?: readonly number[],
  R0?: Matrix,
): { gamma: number[][]; levels: number[] } {
  const { Rm, wv, R0m } = prepareCategoricalInput(R, w, R0);
  const C = wv.length - 1;

  const { levels, inverse } = uniqueLevels(wv);
  const L = levels.length;

  const nCounts = rowBincountWide(Rm, C + 1);
  const n0Counts = rowBincountWide(R0m, C + 1).map((row) => row.map((c) => c + 1));

  const gamma = Rm.map((_, row) => {
    const g = new Array<number>(L).fill(0);
    for (let cat = 0; cat <= C; cat++) {
      g[inverse[cat]!]! += nCounts[row]![cat]! + n0Counts[row]![cat]!;
    }
    return g;
  });

  return { gamma, levels };
}

/**
 * `E[X^k (X+Y)^k]` for `X, Y` from a 3-part Dirichlet partition.
 *
 * `X` has parameter `a`, `Y` has parameter `b`, and the omitted remainder has
 * parameter `total - a - b`. Follows from the multinomial expansion of
 * `(X+Y)^k` and Dirichlet raw moments.
 */
function dirichletNestedCumulativeMoment(
  total: number,
  a: number,
  b: number,
  k: number,
): number {
  if (b <= 0.0) {
    throw new Error("b must be > 0 for nested cumulative moments");
  }
  const logTerms: number[] = [];
  for (let r = 0; r <= k; r++) {
    logTerms.push(
      gammaln(k + 1.0) -
        gammaln(r + 1.0) -
        gammaln(k - r + 1.0) +
        gammaln(a + k + r) -
        gammaln(a) +
        gammaln(b + k - r) -
        gammaln(b) -
        (gammaln(total + 2.0 * k) - gammaln(total)),
    );
  }
  return Math.exp(logSumExp(logTerms));
}

/**
 * Max@k: expected best reward among `k` sampled traces.
 *
 * When `w = [0, 1]`, Max@k reduces exactly to Pass@k. More generally, the
 * reward vector `w` maps categorical outcomes to arbitrary real-valued scores,
 * and Max@k averages the best score obtainable from a subset of size `k`.
 *
 * The finite-sample estimator matches Bagirov et al. (2025), Appendix C.1 /
 * Listing 1 (arXiv:2510.23393).
 *
 * @param R `M x N` categorical outcome matrix with integer entries in
 *          `{0,...,C}`.
 * @param k Number of selected samples, with `1 <= k <= N`.
 * @param w Optional reward vector of shape `(C+1,)`. If omitted, `R` must be
 *          binary and `[0, 1]` is used.
 * @returns Average Max@k score across prompts.
 */
export function maxAtK(R: Matrix, k: number, w?: readonly number[]): number {
  const { Rm, wv } = prepareCategoricalInput(R, w);
  const N = Rm[0]!.length;
  validateK(N, k);

  const denom = comb(N, k);
  // coeff[i - (k-1)] = C(i, k-1) / C(N, k) for i in {k-1, ..., N-1}.
  const coeff: number[] = [];
  for (let i = k - 1; i < N; i++) coeff.push(comb(i, k - 1) / denom);

  let acc = 0;
  for (const row of Rm) {
    const sorted = row.map((c) => wv[c]!).sort((a, b) => a - b);
    let val = 0;
    for (let j = 0; j < coeff.length; j++) {
      val += coeff[j]! * sorted[k - 1 + j]!;
    }
    acc += val;
  }
  return acc / Rm.length;
}

/** Posterior mean/std for Max@k under a grouped Dirichlet posterior. */
function maxAtKBayes(
  R: Matrix,
  k: number,
  w?: readonly number[],
  R0?: Matrix,
): { mu: number; sigma: number; levels: number[] } {
  const { gamma, levels } = groupedPosteriorParams(R, w, R0);
  const M = gamma.length;
  const L = gamma[0]!.length;
  const total = gamma[0]!.reduce((s, v) => s + v, 0);

  if (k < 1 || !Number.isInteger(k)) {
    throw new Error(`k must be an integer >= 1; got ${k}`);
  }
  // The posterior moments describe the latent distribution, so k is not
  // restricted by the observed sample size once the posterior is defined.

  if (L === 1) {
    return { mu: levels[0]!, sigma: 0.0, levels };
  }

  const gaps: number[] = [];
  for (let i = 1; i < L; i++) gaps.push(levels[i]! - levels[i - 1]!);
  const top = levels[L - 1]!;

  const means = new Array<number>(M);
  const vars_ = new Array<number>(M);

  for (let row = 0; row < M; row++) {
    const gammaRow = gamma[row]!;
    // cum[idx] = A_l parameters for l = 1..L-1 (cumulative sums, excluding last).
    const cum: number[] = [];
    let running = 0;
    for (let i = 0; i < L - 1; i++) {
      running += gammaRow[i]!;
      cum.push(running);
    }

    const eAk = new Array<number>(L - 1);
    const eA2k = new Array<number>(L - 1);
    for (let idx = 0; idx < L - 1; idx++) {
      const a = cum[idx]!;
      const b = total - a;
      eAk[idx] = betaRatio(a, b, k, 0);
      eA2k[idx] = betaRatio(a, b, 2 * k, 0);
    }

    let dotGapsEAk = 0;
    for (let i = 0; i < L - 1; i++) dotGapsEAk += gaps[i]! * eAk[i]!;
    const m = top - dotGapsEAk;

    // cross is the (L-1)x(L-1) matrix of cumulative cross moments.
    const cross: number[][] = Array.from({ length: L - 1 }, () =>
      new Array<number>(L - 1).fill(0),
    );
    for (let i = 0; i < L - 1; i++) {
      cross[i]![i] = eA2k[i]!;
      for (let j = i + 1; j < L - 1; j++) {
        const a = cum[i]!;
        const b = cum[j]! - cum[i]!;
        const moment = dirichletNestedCumulativeMoment(total, a, b, k);
        cross[i]![j] = moment;
        cross[j]![i] = moment;
      }
    }

    let e2 = top * top - 2.0 * top * dotGapsEAk;
    // gaps @ cross @ gaps
    let quad = 0;
    for (let i = 0; i < L - 1; i++) {
      let rowDot = 0;
      for (let j = 0; j < L - 1; j++) rowDot += cross[i]![j]! * gaps[j]!;
      quad += gaps[i]! * rowDot;
    }
    e2 += quad;

    means[row] = m;
    vars_[row] = Math.max(0.0, e2 - m * m);
  }

  const mu = means.reduce((s, v) => s + v, 0) / M;
  const sigma = Math.sqrt(vars_.reduce((s, v) => s + v, 0)) / M;
  return { mu, sigma, levels };
}

/**
 * Bayesian posterior summary for {@link maxAtK}, returning `[mu, sigma, lo, hi]`.
 *
 * The posterior uses the same Dirichlet-plus-one construction as `bayes`. When
 * `k = 1`, Max@1 reduces to the single-draw expected score, so this function
 * agrees with `bayesCi`. This uncertainty model is a `scorio` extension and is
 * not part of Bagirov et al. (2025).
 *
 * @param R `M x N` categorical outcome matrix with integer entries in
 *          `{0,...,C}`.
 * @param k Selection count; defined for any integer `k >= 1`. `k = 1` matches
 *          `bayesCi`.
 * @param w Optional reward vector of shape `(C+1,)`. If omitted, `R` must be
 *          binary and `[0, 1]` is used.
 * @param R0 Optional `M x D` matrix of prior outcomes.
 * @param confidence Credibility level for the normal-approximation interval.
 * @param bounds Optional `[lo, hi]` clipping bounds. If omitted, the interval
 *          is clipped to the minimum and maximum reward levels in `w`.
 * @returns `[mu, sigma, lo, hi]`.
 */
export function maxAtKCi(
  R: Matrix,
  k: number,
  w?: readonly number[],
  R0?: Matrix,
  confidence = 0.95,
  bounds?: Bounds,
): [number, number, number, number] {
  if (k === 1) {
    return bayesCi(R, w, R0, confidence, bounds);
  }

  const { mu, sigma, levels } = maxAtKBayes(R, k, w, R0);
  const effectiveBounds: Bounds =
    bounds === undefined
      ? [Math.min(...levels), Math.max(...levels)]
      : bounds;
  const [lo, hi] = normalCredibleInterval(mu, sigma, confidence, true, effectiveBounds);
  return [mu, sigma, lo, hi];
}
